File: leetcode/101-200/153-find-minimum-in-rotated-sorted-array/linear_scan.py
"""LeetCode #153: Find Minimum in Rotated Sorted Array — linear scan O(n).

Algorithmic mirror of linear_scan.kara. Output format matches line-for-line
so the two can be diffed directly.
"""

from __future__ import annotations


def find_min(nums: list[int]) -> int:
    m = nums[0]
    for i in range(1, len(nums)):
        x = nums[i]
        if x < m:
            m = x
    return m


def report(nums: list[int]) -> None:
    print(find_min(nums))


def main() -> None:
    report([3, 4, 5, 1, 2])         # expect: 1
    report([4, 5, 6, 7, 0, 1, 2])   # expect: 0
    report([11, 13, 15, 17])        # expect: 11
    report([5])                     # expect: 5
    report([2, 1])                  # expect: 1


if __name__ == "__main__":
    main()

File: leetcode/101-200/153-find-minimum-in-rotated-sorted-array/bench/linear_scan.c
/*
 * LeetCode 153 — linear-scan find-min on a rotated sorted array, bench mirror in C.
 *
 * Algorithmic mirror of bench/linear_scan.{kara,rs,py}. N=2_000_000,
 * pivot R=666_666, K=10 outer iterations. Stdout sink: 10.
 *
 * See ../README.md § Benchmarks for what the numbers mean.
 */
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>

#define N 2000000
#define R 666666
#define K 10

static const int64_t *black_box_slice(const int64_t *p) {
#if defined(__GNUC__) || defined(__clang__)
    __asm__ volatile("" : "+r"(p) :: "memory");
#endif
    return p;
}

static int64_t find_min(const int64_t *nums, size_t len) {
    int64_t m = nums[0];
    for (size_t i = 1; i < len; i++) {
        if (nums[i] < m) {
            m = nums[i];
        }
    }
    return m;
}

int main(void) {
    int64_t *data = (int64_t *)malloc((size_t)N * sizeof(int64_t));
    if (!data) {
        fprintf(stderr, "malloc failed\n");
        return 1;
    }
    for (size_t i = 0; i < (size_t)N; i++) {
        data[i] = (int64_t)(((i + R) % N) + 1);
    }

    int64_t sum = 0;
    for (int k = 0; k < K; k++) {
        sum += find_min(black_box_slice(data), (size_t)N);
    }
    printf("%lld\n", (long long)sum);
    free(data);
    return 0;
}
